packages/controls/src/controls/font/font.ts

```diff
@@ -0,0 +1,318 @@
+          withVariants.optional(),
+          versionedDataWithVariants.optional(),
+          z.literal<true>(true),
+        ),
+        strict: schemas(
+          withVariants,
+          versionedDataWithVariants,
+          z.literal<true>(true),
+        ),
+      },
+    }
+  }
+
+  static deserialize(data: DeserializedRecord): FontDefinition {
+    if (data.type !== Definition.type) {
+      throw new Error(
+        `Font: expected type ${Definition.type}, got ${data.type}`,
+      )
+    }
+
+    const { config, version } = z
+      .union([
+        Definition.schema.withVariants.relaxed.definition,
+        Definition.schema.withoutVariants.relaxed.definition,
+      ])
+      .parse(data)
+    return new FontDefinition(config, version)
+  }
+
+  constructor(
+    config: C,
+    readonly version: z.infer<typeof Definition.schema.version>,
+  ) {
+    super(config)
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema(): ReturnedSchemaType<C> {
+    const { withVariants, withoutVariants, ...baseSchema } = Definition.schema
+
+    const refinedSchema = this.refinedSchema
+
+    const schema = {
+      ...baseSchema,
+      type: refinedSchema.type,
+      definition: refinedSchema.definition,
+      value: refinedSchema.value,
+      data: refinedSchema.data,
+      resolvedValue: refinedSchema.resolvedValue,
+    }
+
+    return schema as ReturnedSchemaType<C>
+  }
+
+  get refinedSchema() {
+    return this.config.variant === false
+      ? this.config.defaultValue === undefined
+        ? Definition.schema.withoutVariants.relaxed
+        : Definition.schema.withoutVariants.strict
+      : this.config.defaultValue === undefined
+        ? Definition.schema.withVariants.relaxed
+        : Definition.schema.withVariants.strict
+  }
+
+  get dataSchema(): Schema<C>['data'] {
+    return this.refinedSchema.data
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
+    return safeParse(this.dataSchema, data)
+  }
+
+  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
+    const inputSchema = this.dataSchema.optional()
+    return match(data satisfies z.infer<typeof inputSchema>)
+      .with(Definition.dataSignature.v1, ({ value }) => value)
+      .otherwise(() => undefined)
+  }
+
+  toData(value: ValueType<C>): DataType<C> {
+    if (value == null) return undefined
+
+    return {
+      ...Definition.dataSignature.v1,
+      value,
+    }
+  }
+
+  copyData(data: DataType<C> | undefined): DataType<C> | undefined {
+    if (data == null) return data
+
+    return {
+      ...data,
+      value: {
+        ...data.value,
+      },
+    }
+  }
+
+  resolveValue(
+    data: DataType<C> | undefined,
+  ): Resolvable<ResolvedValueType<C> | undefined> {
+    return {
+      readStableValue: (_previous?: ResolvedValueType<C>) => {
+        return (
+          this.fromData(data) ??
+          (this.config.defaultValue as ResolvedValueType<C>)
```

Comment:
Really glad I wrote those tests because I was missing this @migueloller!



Here is the diff of me addressing your comments: https://github.com/makeswift/makeswift/compare/7d4b593a4a6dc6cc3b34a79f51e361f5d0d67ee8..4787c00965154e422cfb4c78e6aec7ca759fd20c

Reply:
Nice